common/splice_script.c

```diff
@@ -0,0 +1,2696 @@
+
+				tokens[n++] = new_token(tokens, TOK_ATSYM,
+							script_index);
+
+				script_index++;
+
+				tokens[n] = new_token(tokens, TOK_STR,
+						      script_index);
+				tokens[n++]->str = split_point + 1;
+			} else {
+				tokens[n++] = tal_steal(tokens, input[i]);
+			}
+			break;
+		case TOK_ATSYM:
+		case TOK_SATS:
+		case TOK_PERCENT:
+		case TOK_QUESTION:
+		case TOK_WILDCARD:
+		case TOK_FEE:
+		case TOK_CHANID:
+		case TOK_WALLET:
+		case TOK_FEERATE:
+		case TOK_NODEID:
+		case TOK_BTCADDR:
+		case TOK_CHANQUERY:
+		case TOK_MULTI_CHANID:
+		case TOK_LEASERATE:
+		case TOK_LEASEREQ:
+		case TOK_SEGMENT:
+			return new_error(ctx, INVALID_TOKEN, input[i],
+					 "3rd_separators");
+		}
+	}
+
+	tal_free(input);
+	tal_resize(&tokens, n);
+	*tokens_inout = tokens;
+	return NULL;
+}
+
+static const char *segwit_addr_net_decode(int *witness_version,
+					  uint8_t *witness_program,
+					  size_t *witness_program_len,
+					  const char *addrz,
+					  const struct chainparams *chainparams)
+{
+	if (segwit_addr_decode(witness_version, witness_program,
+			       witness_program_len, chainparams->onchain_hrp,
+			       addrz))
+		return chainparams->onchain_hrp;
+	else
+		return NULL;
+}
+
+static bool is_bitcoin_address(const char *address)
+{
+	struct ripemd160 addr;
+	int witness_version;
+	/* segwit_addr_net_decode requires a buffer of size 40, and will
+	 * not write to the buffer if the address is too long, so a buffer
+	 * of fixed size 40 will not overflow. */
+	uint8_t witness_program[40];
+	size_t witness_program_len;
+
+	const char *bech32;
+
+	u8 addr_version;
+
+	if (ripemd160_from_base58(&addr_version, &addr,
+				  address, strlen(address))) {
+		if (addr_version == chainparams->p2pkh_version) {
+			return true;
+		} else if (addr_version == chainparams->p2sh_version) {
+			return true;
+		}
+		return false;
+	}
+
+	bech32 = segwit_addr_net_decode(&witness_version, witness_program,
+					&witness_program_len, address,
+					chainparams);
+	if (bech32) {
+		bool witness_ok;
+
+		/* Only V0 has restricted lengths of witness programs */
+		if (witness_version == 0) {
+			witness_ok = (witness_program_len == 20 ||
+				       witness_program_len == 32);
+		} else
+			witness_ok = true;
+
+		if (!witness_ok)
+			return false;
+
+		if (!streq(bech32, chainparams->onchain_hrp))
+			return false;
+
+		return true;
+	}
+
+	return false;
+}
+
+/* Checks token->str for a short node id and auto completes it. */
+static bool autocomplete_node_id(struct token *token,
+				 struct splice_script_chan *channels,
+				 size_t channels_count,
+				 bool *multiple_nodes,
+				 bool *chan_id_overmatch)
+{
+	struct node_id *match;
+	struct node_id candidate;
+	size_t len = strlen(token->str) / 2;
+
+	*multiple_nodes = false;
+	*chan_id_overmatch = false;
+
+	if (strlen(token->str) < NODEID_MIN_CHARS)
+		return false;
+	if (len > PUBKEY_CMPR_LEN)
+		return false;
+	if (!hex_decode(token->str, len * 2,
```

Comment:
Did you want to match half-bytes here?

I'm not sure if an odd hex digit is the upper half of the byte or the lower half. I think it may not be defined

Most of the internet seems to just reject odd number hex strings. Google addresses odd number of hex digits by prepending an extra 0 on the front.
![IMAGE 2024-05-30 16:24:01](https://github.com/ElementsProject/lightning/assets/331415/bc20f33e-f5e0-4289-ad3f-bb8eb774d4e7)
![IMAGE 2024-05-30 16:24:09](https://github.com/ElementsProject/lightning/assets/331415/ed86fa3b-6ad4-4561-92c7-47657490ea50)

Which would definetely *not* work well for channel identifiers

If I have a channel
`0399069f1693fd89a453f0caf03ee36b6f6c8abaa7ef778d3e2bcc7c2b44120100`
And I querty it with
`03990`
I definetely do not mean to search for 
`003990`

So we'd have to invent a new spec for hex encoding which would probably go against consensus of how they are used.

I prefer dropping the rightmost digit for odd numbered hex strings as then `03990` would match our example.